### PR TITLE
Ability to invalidate cache items on request 

### LINKED ISSUE #24: Support for cache skip or invalidation
In my application, I need to be able to force delete whichever cache I had set. So, I would like to suggest the ability to be able to delete a cached key, without the need to perform the request itself. That is useful when the developer knows that some data is stale, but does not need to request it yet.

In fact, it probably is better to have the option to set the data stale, rather than deleting it. In which case #20 would still work as intended, but a new request would ignore the stale status.

Additionally, it would be useful, if on a request basis, the developer would be able to state that the cache should be ignored if it exists, but the response should be cached when retrieved. Presently, `excludeFromCache` does not allow for ignoring the cached version while overriding the existing one. This is likely to require #23.

Lastly, but that should probably make its way to another issue, my implementation needs to be able to invalidate cache in bulk. I don't know if this should be the responsibility of your adapter, but my use case is that when some data is set, I know that a bunch of other queries' cache has to be marked stale. For example, when the user creates a new todo, I know that the list of todos is invalid. But also the timeline showing the recent todos, and perhaps something else. So, I am marking the request to be part of some groups, and I allow the cache to be invalidated by group name. E.g.:

```
// Prior to requesting the data.

if (!isExcluded && req.cache && req.cache.groups) {
  const groupsCacheKey = '__groups';
  const groupsKeys = (await cacheStore.getItem(groupsCacheKey)) || {};
  let hasSetAny = false;

  for (let group of req.cache.groups) {
    if (!(group in groupsKeys)) {
      groupsKeys[group] = [];
    }
    if (groupsKeys[group].indexOf(uuid) < 0) {
      hasSetAny = true;
      groupsKeys[group].push(uuid);
    }
  }

  // Commit the changes.
  if (hasSetAny) {
    await cacheStore.setItem(groupsCacheKey, groupsKeys);
  }
}

// Perform the request...
```

```
// Cache clearing functions.
const clearCacheByKey = async function(uuid) {
  let result = await cacheStore.getItem(uuid);
  if (result && 'expires' in result) {
    result.expires = 1;
    await cacheStore.setItem(uuid, result);
  }
};

const clearCacheByGroup = async function(group) {
  const groups = (await cacheStore.getItem('__groups')) || {};
  const uuids = groups[group] || [];
  for (let uuid of uuids) {
    await clearCacheByKey(uuid);
  }
};
```

The actions modifying the list of groups are not atomic, which can be an issue, but for now I'm OK to live with it.

### PR DESCRIPTION
**Pull request by jahead**
Fixes #24

### CODE REVIEW

**Comment by RasCarlito** [comment]
Hey @jahead 😃 

Tested this out a bit and I really like it. I would like to get this merged 🚀 

Just a few points:
* I think we should add some documentation in the readme
* We should add a test where we override the default invalidation method
* Small suggestions on the code

Else it's really neat!

If you'd like I can work with you on this, just tell me.

If we get this done and include a better default URL serializer I can at last make a release with a few notable and useful changes 😄 🎉

File: src/cache.js [hunk 1]
```diff
@@ -74,5 +74,17 @@ function key (config) {
   return cacheKey
 }
 
-export { read, write, key }
-export default { read, write, key }
+function invalidate (config = {}) {
+  if (isFunction(config.invalidate)) return config.invalidate
+
+  return async (cfg, req) => {
+    const uuid = cfg.key(req)
```
Comment (RasCarlito) [inline src/cache.js]: Here you don't need to compute the `uuid` from the `req`. It has already been done [in the `request` method](https://github.com/RasCarlito/axios-cache-adapter/pull/54/files#diff-aab1e3f14ac4fa89a7ece88d240d7a6eL8) and added to the `config` object.

So when we want to remove the item we can do
```js
await cfg.store.removeItem(cfg.uuid)
```

File: src/cache.js [hunk 2]
```diff
@@ -74,5 +74,17 @@ function key (config) {
   return cacheKey
 }
 
-export { read, write, key }
-export default { read, write, key }
+function invalidate (config = {}) {
+  if (isFunction(config.invalidate)) return config.invalidate
+
+  return async (cfg, req) => {
```
Comment (RasCarlito) [inline src/cache.js]: Maybe we could extract this function and name it `defaultInvalidate`. It does not use any closures so it does not need to be defined inside the `invalidate` function.

Something like
```js
function invalidate (config = {}) {
  if (isFunction(config.invalidate)) return config.invalidate

  return defaultInvalidate
}

async function defaultInvalidate (config, req) {
  // ...
}
```

Comment (jahead) [reply src/cache.js]: Ah, yes. That was just a case of lazy programming, should we extract the invalidate related code into a seperate file like you do for other similar features? 

I don't mind, it's not a lot of work to refactor it.

Comment (RasCarlito) [reply src/cache.js]: I was thinking about it but came to the conclusion that it goes well in the `cache.js` file 😄

**Comment by RasCarlito** [comment]
@jahead Please rebase onto `master` branch as some changes have been merged 🚀

File: src/cache.js (hunk 1 above)
Comment (jahead) [reply src/cache.js]: Fixed

File: src/cache.js (hunk 2 above)
Comment (jahead) [reply src/cache.js]: Fixed

**Comment by RasCarlito** [comment]
@jahead Awesome 🎉 

When we get this merged I'll release the v2.2.0

Do you want me to update the docs or are you already working on something?

Cheers 🍻

**Comment by jahead** [comment]
Okay, I just made a test for this. 

I could write a little doco, but probably not tonight. I see if I can do it tomorrow. 

Just a little OT but you define your memory store iterate function as (value, key) as opposed to (key, value). 

Was just wondering what the reasoning is as I wrote an AsyncStorage wrapper for react native's storage subsystem, and it assumes (key, value) because I wasn't aware haha.

**Comment by RasCarlito** [comment]
Test and code looks good 👍 

If you don't mind I can do the docs today (morning here in france 😄) so that I can get this released at the same time. What do you think?

The store API is the same as [`localforage`](https://localforage.github.io/localForage/#data-api-iterate). The adapter was designed with `localforage` in mind. The "in house" memory store was just made in case people don't want to add another dependency and to make the adapter work in Node.js

I'm planning a v3 maybe this store could evolve into something else then.

Have a good night!

**Comment by RasCarlito** [comment]
After thinking about it a bit more, now I don't see any problem with a `HEAD` request invalidating cache 😅 

And moving cache invalidation after cache exclusion logic does not work.

**Comment by RasCarlito** [comment]
@jahead Would you have anything else to add to this PR?

If not I'll get this merged and make a new release 🎉

**Comment by jahead** [comment]
Yea, sorry. I only just got to this. But I'm of the same view.
Ideally Invalidate could happen in all requests. And in response, given the what comes back from a head request for example.

Though invalidation based on response should be handled in different PR. 
Nothing else to add.

**Pull request merged**